### PR TITLE
#165305262 Email verification on sign up

### PR DESCRIPTION
**Pull request by AlvinMugambi**
### What does this PR do?

Implement email verification on user sign up

### Description of the task to be completed?

- Ensure an email is sent to a user once they sign up in order to activate their account
- Refactor and add tests
- Update environment variables on the env-example
- update requirements.txt

### How should you manually test this?

- [x] **Clone the repository**

        $ git clone https://github.com/andela/ah-the-jedi-backend.git

- [x] **Switch to testing branch**

       $ git checkout ft-email-verification-165305262

- [x] **Switch to the ah-the-jedi-backend directory**

       $ cd ah-the-jedi-backend/

 

- [x] **Create and activate the virtual environment**

      $  virtualenv -p python .venv

      $ source .venv/bin/activate

- [x] **Install project dependencies**

       $ pip install -r requirements.txt

- [x] **Run the database migrations**

       $ python manage.py migrate

### Running the application

      $ python manage.py runserver

- [x] **Signup a user**

           POST <your-localhost>/api/users/

- [x] **Go to the registered email inbox and click verification link**

- [x] **Copy link url to postman**

       POST <your-localhost>/api/users/activate/?uid={uid}&token={token} 

### Testing

      $ python manage.py test

### Prerequisites

Python, Git,  Virtualenv

### What are the relevant PT stories?

[#165305262](https://www.pivotaltracker.com/story/show/165305262)

### Screenshots
*Sign up user*
Screenshot 2019-04-29 at 12 38 58

*Check email*
Screenshot 2019-04-26 at 03 00 37
*Click on verify email*
Screenshot 2019-04-26 at 03 00 45
*Copy link to postman and post*
Screenshot 2019-04-26 at 03 01 15

### CODE REVIEW

File: authors/apps/authentication/views.py [hunk 1]
```diff
@@ -80,3 +85,34 @@ def update(self, request, *args, **kwargs):
...
+class ActivationView(generics.GenericAPIView):
+    """
+    post:
+    Activate user.
+    """
+    serializer_class = UidAndTokenSerializer
+    permission_classes = [permissions.AllowAny]
+    token_generator = default_token_generator
+
+    def post(self, request):
```
Comment (katherine95) [inline authors/apps/authentication/views.py:104]: LGTM. Great work in ensuring the `uid` and `token` fields are fetched from the query params.

**Review by katherine95 — APPROVED** [review]

**Comment by ElishaMisoi** [comment]
LGTM 👍🏽

**Review by ElishaMisoi — APPROVED** [review]

File: authors/apps/authentication/urls.py [hunk 2]
```diff
@@ -1,15 +1,18 @@
...
+    LoginAPIView, RegistrationAPIView, UserRetrieveUpdateAPIView, ActivationView
 )
 
 swagger_view = get_swagger_view(title='The Jedi Authors Haven API')
 
 urlpatterns = [
-    url(r'^user/?$', UserRetrieveUpdateAPIView.as_view()),
-    url(r'^users/?$', RegistrationAPIView.as_view()),
-    url(r'^users/login/?$', LoginAPIView.as_view()),
+    path('user/', UserRetrieveUpdateAPIView.as_view()),
```
Comment (Benkimeric) [inline authors/apps/authentication/urls.py]: Looks good to me. Nice job in implementing the use of `path` instead of `url` in the application.

**Review by jkamz — APPROVED** [review]
LGTM. Really quality code. Keep it up.

**Review by Benkimeric — APPROVED** [review]
Looks good to me. Nice job in implementing the use of `path` instead of `url` in the application.

**Review by Siffersari — APPROVED** [review]
LGTM. Good implementation and formatting the tests and test data.

**Review by rnjane — COMMENTED** [review]
1. Add a message to the returned response upon email verification indicating email has been successfully verified.
2. Add email setup instructions in the readme to ease local setup for others.

**Comment by AlvinMugambi** [comment]
> 1. Add a message to the returned response upon email verification indicating email has been successfully verified.
> 2. Add email setup instructions in the readme to ease local setup for others.

**Comment by AlvinMugambi** [comment]
> 1. Add a message to the returned response upon email verification indicating email has been successfully verified.
> 2. Add email setup instructions in the readme to ease local setup for others.

Successfully made the changes @rnjane

**Review by rnjane — APPROVED** [review]

**Pull request merged**